docs/dag-sharding-rfc.md

```diff
@@ -0,0 +1,182 @@
+
+If ipfs-cluster is to stream DAG nodes from its local daemon to create a Cluster DAG, the ipfs api would need to include an endpont that provides a stream of a DAG's blocks without overcommitting the daemon's resources (i.e. not downloading the whole DAG first if the repo is too small).  One relevant effort is the [CAR (Certified ARchives) project](https://github.com/ipld/specs/pull/51) which aims to define a format for storing DAGs.  One of its goals as stated in a recent proposal is to explicitly allow users to "Traverse a large DAG streamed over HTTP without downloading the entire thing."  Cluster could shard DAG nodes extracted from a streamed CAR file, either kept on disk or provided at a location on the network.  CAR integration with go-ipfs could potentially allow resource aware streaming of DAG nodes over the ipfs daemon's api so that the sharding cluster peer need only know the DAG's root hash.
+
+Another relevant effort is that of ipld selectors.  ipld selectors are interesting because they are an orthagonal approach to the Cluster DAG for succinctly organizing large pinsets.  We did not end up including them in the main proposal because adding large datasets necessitates importing data to a DAG anyway.  In this case, from our perspective, building a management DAG for pinning and directly transferring blocks is simpler than adding a portion of the DAG to the local ipfs node, executing a remote selector pin, blocking on ipfs to do the transfer, and then clearing resources from the local node.  However in the case that the dag is already on ipfs, pinning selectors for shards is a potential implementation strategy.  The selectors would essentially act as the shard-nodes of the Cluster DAG, selecting shards of the DAG that could fit on individual nodes and getting pinned by the assigned nodes.  At the very least this seems to require that ipld selectors can select based on sub-tree size.  It is currently unclear how well matched ipld selectors would be for this use case.
+
+# Workplan
+
+There is a fair amount that needs to be built, updated and serviced to make all of this work.  Here I am focusing on basic sharding, allowing support for files that wouldn't fit in an ipfs repo.  Support for fault tolerance, more clever sharding strategies and sharding large existing DAGs will come after this is nailed down.  
+
+* Build up a version of the `importers` (aka `dagger` aka `dex`) project that suits cluster's needs (if possible keep it general enough for general use).  In short we want a library that imports data to ipfs DAGs from arbitrary formats (tar, deb, unix files, etc) with arbitrary chunking and layout algorithms.  For our use case it is important that the interfaces support taking in a stream of data and returning a stream of blocks or DAG nodes.
+* Implement a cluster importer component.  It will need to provide endpoints that supporting streams as input and output.
```

Comment:
probably this can be removed as `dex` will do the job. and the rest api will use it directly.